src/Components/Server/src/ComponentHub.cs

```diff
@@ -220,6 +238,157 @@ public async ValueTask<bool> ConnectCircuit(string circuitIdSecret)
+    //     two separate server instances (for example, server A, B, where it starts resuming on A, something fails and tries to start resuming on B)
+    //     the state would either be ignored in one case or lost.
+    //   * Two things can happen:
+    //     * Both A and B are somehow able to read the same state.
+    //       * Even if A gets the state, it doesn't complete the "resume" handshake, so its state gets discarded
+    //       and not saved again.
+    //       * B might complete the handshake and then the circuit will resume on B.
+    //     * A deletes the state before B is able to read it. Then "resumption" fails, as the circuit state is gone.
+
+    // On the server we are going to have a public method on Circuit.cs to trigger pausing a circuit from the server
+    // that returns the root components and application state as strings data-protected by the data protection provider.
+    // Those can be then passed to this method for resuming the circuit.
+    public async ValueTask<string> ResumeCircuit(
+        string circuitIdSecret,
+        string baseUri,
+        string uri,
+        string rootComponents,
+        string applicationState)
+    {
+        // TryParseCircuitId will not throw.
+        if (!_circuitIdFactory.TryParseCircuitId(circuitIdSecret, out var circuitId))
+        {
+            // Invalid id.
+            Log.ResumeInvalidCircuitId(_logger, circuitIdSecret);
+            return null;
+        }
+
+        var circuitHost = _circuitHandleRegistry.GetCircuit(Context.Items, CircuitKey);
+        if (circuitHost != null)
+        {
+            // This is an error condition and an attempt to bind multiple circuits to a single connection.
+            // We can reject this and terminate the connection.
+            Log.CircuitAlreadyInitialized(_logger, circuitHost.CircuitId);
+            await NotifyClientError(Clients.Caller, $"The circuit host '{circuitHost.CircuitId}' has already been initialized.");
+            Context.Abort();
+            return null;
+        }
+
+        if (baseUri == null ||
+            uri == null ||
+            !Uri.TryCreate(baseUri, UriKind.Absolute, out _) ||
+            !Uri.TryCreate(uri, UriKind.Absolute, out _))
+        {
+            // We do some really minimal validation here to prevent obviously wrong data from getting in
+            // without duplicating too much logic.
+            //
+            // This is an error condition attempting to initialize the circuit in a way that would fail.
+            // We can reject this and terminate the connection.
+            Log.InvalidInputData(_logger);
+            await NotifyClientError(Clients.Caller, "The uris provided are invalid.");
+            Context.Abort();
+            return null;
+        }
+
+        PersistedCircuitState? persistedCircuitState;
+        if (rootComponents == "[]" && string.IsNullOrEmpty(applicationState))
```

Comment:
Should this line also be using `RootComponentIsEmpty()` like the `else if` block below it?